### PR TITLE
Add progressive web app support

### PR DESCRIPTION
**Pull request by aik9508**
Add interfaces of serviceWorker API and Promise API which allow users to create progressive web apps with Ocsigen. Several necessary functions are also added in Url.ml and Form.ml.

### CODE REVIEW

**Comment by tbrk** [comment]
It would be great to have an Lwt wrapper for Promises (I'm currently trying to wrap a Javascript library that uses them). Would there be more chance of acceptance if the promises part were separated from the other proposed changes?

**Comment by hhugo** [comment]
Splitting this PR in pieces would definitely help. Regarding the lwt wrapper for Promises, I'd like to have input from @aantron (who wrote https://github.com/aantron/repromise)

File: lib/promise.ml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  Js.Unsafe.meth_call _Promise "resolve" [|Js.Unsafe.inject v|]
+
+let resolve_promise (v :'a promise Js.t) : 'a promise Js.t =
+  Js.Unsafe.meth_call _Promise "resolve" [|Js.Unsafe.inject v|]
+
+let reject_value (v :'a) : 'a promise Js.t =
+  Js.Unsafe.meth_call _Promise "reject" [|Js.Unsafe.inject v|]
+
+let reject_promise (v :'a promise Js.t) : 'a promise Js.t =
+  Js.Unsafe.meth_call _Promise "rejetct" [|Js.Unsafe.inject v|]
```
Comment (tbrk) [inline lib/promise.ml]: This looks like a typo to me: _rejetct_.

**Comment by tbrk** [comment]
Thanks @aantron for such detailed and useful remarks. I had no idea that Javascript promises were so problematic to wrap.

**Comment by hhugo** [comment]
Closing this PR. Please reopen as separate (multiple) PRs if still needed.

**Pull request closed without merging**